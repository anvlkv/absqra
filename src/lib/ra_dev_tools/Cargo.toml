[package]
name = "ra_dev_tools"
version = "0.1.0"
authors = ["a.nvlkv <a.nvlkv@protonmail.ch>"]
edition = "2018"
build = "build.rs"

[build-dependencies]
build-deps = "0.1.4"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = "0.16.0"
regex = "1.3.9"
example_files_macro = {path="./src/lib/example_files_macro"}
files_iterator = {path="./src/lib/files_iterator"}
