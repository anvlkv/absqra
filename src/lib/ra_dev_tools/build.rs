extern crate build_deps;

fn main() {
    // Enumerate files in sub-folder "examples/*", being relevant for the test-generation (as example)
    // If function returns with error, exit with error message.
    build_deps::rerun_if_changed_paths( "../../../examples/*/*" ).unwrap();

    // Adding the parent directory "examples" to the watch-list will capture new-files being added
    build_deps::rerun_if_changed_paths( "../../../examples/*" ).unwrap();
}