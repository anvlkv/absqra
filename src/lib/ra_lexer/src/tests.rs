use crate::cursor::{Cursor, Position};
use crate::{tokenize, RaToken, TokenKind};

mod cursor {
    use super::{Cursor, Position};
    #[test]
    fn it_should_create() {
        Cursor::new("abc", Position(1, 0), 1, 0);
    }

    #[test]
    fn it_should_give_first_char() {
        let cur = Cursor::new("abc", Position(1, 0), 1, 0);
        assert_eq!(cur.first_ahead(), 'a');
    }

    #[test]
    fn it_should_give_second_char() {
        let cur = Cursor::new("abc", Position(1, 0), 1, 0);
        assert_eq!(cur.second_ahead(), 'b');
    }

    #[test]
    fn it_should_return_next_char() {
        let mut cur = Cursor::new("a", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
    }

    #[test]
    fn it_should_return_none_at_end_of_input() {
        let mut cur = Cursor::new("a", Position(1, 0), 1, 0);
        cur.bump();
        assert_eq!(cur.bump(), None);
    }

    #[test]
    fn it_should_confirm_eof() {
        let mut cur = Cursor::new("a", Position(1, 0), 1, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.is_eof(), true);
    }

    #[test]
    fn it_should_return_amount_of_consumed_symbols() {
        let mut cur = Cursor::new("abc", Position(1, 0), 1, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.len_consumed(), 3);
    }

    #[test]
    fn it_should_not_panic_when_encountering_funny_characters() {
        let mut cur = Cursor::new("ðŸš¬", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'ðŸš¬');
    }

    #[test]
    fn it_should_track_position() {
        let mut cur = Cursor::new("abc\nSOME", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(1, 1));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(1, 2));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(1, 3));
        assert_eq!(cur.bump().unwrap(), 'S');
        assert_eq!(cur.position, Position(2, 1));
        assert_eq!(cur.bump().unwrap(), 'O');
        assert_eq!(cur.position, Position(2, 2));
        assert_eq!(cur.bump().unwrap(), 'M');
        assert_eq!(cur.position, Position(2, 3));
    }

    #[test]
    fn it_should_track_indent_level() {
        let mut cur = Cursor::new("a\n\tb\n\t\tc\nd", Position(1, 0), 1, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.level, 1);
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.level, 2);
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.level, 3);
        assert_eq!(cur.position, Position(3, 3));
        assert_eq!(cur.bump().unwrap(), 'd');
        assert_eq!(cur.level, 1);
    }
}

mod lib {
    use super::{tokenize, Position, RaToken, TokenKind};

    #[test]
    fn it_should_create_iterator_of_tokens() {
        let mut stream = tokenize("abc");
        stream.next();
    }

    #[test]
    fn it_should_parse_multiple_lines() {
        let mut stream = tokenize("!?&/\n\tabc");
        stream.next();
        stream.next();
        stream.next();
        stream.next();
        assert_eq!(
            stream.next().unwrap().unwrap(),
            RaToken {
                kind: TokenKind::Identifier("abc"),
                content: "abc",
                position: (Position(2, 1), Position(2, 4)),
                len: 3,
                level: 2,
            }
        )
    }

    #[test]
    fn it_should_parse_identifiers() {
        let mut stream = tokenize("abc");
        assert_eq!(
            stream.next().unwrap().unwrap(),
            RaToken {
                kind: TokenKind::Identifier("abc"),
                content: "abc",
                position: (Position(1, 0), Position(1, 3)),
                len: 3,
                level: 1
            }
        );
    }

    use ra_dev_tools::insta::assert_json_snapshot;
    use ra_dev_tools::make_example_tests;

    #[make_example_tests]
    #[test]
    fn it_should_match_snapshots(contents: String, file_name: String) {
        let tokens: Vec<RaToken> = tokenize(&contents).map(|r| r.unwrap()).collect();
        assert_json_snapshot!(file_name, tokens)
    }

    mod strings {
        use super::{tokenize, Position, RaToken, TokenKind};
        #[test]
        fn it_should_parse_string_literals_with_double_quotes() {
            let mut stream = tokenize("\"some\"");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::StringLiteral("some"),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    level: 1
                }
            )
        }

        #[test]
        fn it_should_parse_string_literals_with_single_quotes() {
            let mut stream = tokenize("\'some\'");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::StringLiteral("some"),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    level: 1
                }
            )
        }
    }
    mod comments {
        use super::{tokenize, Position, RaToken, TokenKind};
        #[test]
        fn it_should_parse_single_line_comments() {
            let mut stream = tokenize("//abc");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 1
                }
            );
        }
        #[test]
        fn it_should_parse_multi_line_comments() {
            let mut stream = tokenize("/*abc\nSOME*/");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "abc\nSOME",
                    position: (Position(1, 0), Position(2, 6)),
                    len: 12,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_parse_multi_line_comments_with_fringe() {
            let mut stream = tokenize("/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "\n* abc\n* SOME\n",
                    position: (Position(1, 0), Position(4, 2)),
                    len: 18,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_parse_comments_followed_by_one_another() {
            let mut stream = tokenize("//some\n\n/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "some",
                    position: (Position(1, 0), Position(1, 6)),
                    len: 6,
                    level: 1
                }
            );

            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Comment,
                    content: "\n* abc\n* SOME\n",
                    position: (Position(3, 0), Position(6, 2)),
                    len: 18,
                    level: 1
                }
            );
        }
    }
    
    mod numbers {
        use super::{tokenize, Position, RaToken, TokenKind};
        use crate::errors::LexerError;
        #[test]
        fn it_should_parse_numbers() {
            let mut stream = tokenize("123");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Int(123),
                    content: "123",
                    position: (Position(1, 0), Position(1, 3)),
                    len: 3,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Int(v) => assert_eq!(v, 123),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_separator() {
            let mut stream = tokenize("123,321");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123.321),
                    content: "123,321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123.321),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_separator() {
            let mut stream = tokenize("123.321");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123.321),
                    content: "123.321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123.321),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_thousands_separator() {
            let mut stream = tokenize("123.321,456");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123321.456),
                    content: "123.321,456",
                    position: (Position(1, 0), Position(1, 11)),
                    len: 11,
                    level: 1
                }
            );

            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123321.456),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_multiple_thousands_separator() {
            let mut stream = tokenize("123.321.123,456");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123321123.456),
                    content: "123.321.123,456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 1
                }
            );

            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123321123.456),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_and_multiple_thousands_separators() {
            let mut stream = tokenize("123,321,123.456");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(123321123.456),
                    content: "123,321,123.456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 1
                }
            );

            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, 123321123.456),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_error_when_encountering_multiple_decimal_separators() {
            let mut stream = tokenize("123.321.123,456,654");
            match stream.next().unwrap() {
                Ok(_) => panic!("didn't error"),
                Err(e) => assert!(match e {
                    LexerError::UnexpectedCharacter(ch, _) => ch == ',',
                    _ => false
                })
            }
        }

        #[test]
        fn it_should_parse_negative_integer() {
            let mut stream = tokenize("-123");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Int(-123),
                    content: "-123",
                    position: (Position(1, 0), Position(1, 4)),
                    len: 4,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Int(v) => assert_eq!(v, -123),
                _ => panic!("unexpected kind")
            };
        }

        #[test]
        fn it_should_parse_negative_float() {
            let mut stream = tokenize("-123.312");
            let parsed = stream.next().unwrap().unwrap();
            assert_eq!(
                parsed,
                RaToken {
                    kind: TokenKind::Float(-123.312),
                    content: "-123.312",
                    position: (Position(1, 0), Position(1, 8)),
                    len: 8,
                    level: 1
                }
            );
            match parsed.kind {
                TokenKind::Float(v) => assert_eq!(v, -123.312),
                _ => panic!("unexpected kind")
            };
        }
    }

    mod content_block {
        use super::{tokenize, Position, RaToken, TokenKind};
        use crate::errors::LexerError;

        #[test]
        fn it_should_parse_content_blocks() {
            let mut stream = tokenize("`abc`");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::ContentBlock,
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_keep_inner_indents_when_parsing_content() {
            let mut stream = tokenize("`\n\tabc\n\t\tabc\n`");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::ContentBlock,
                    content: "\n\tabc\n\t\tabc\n",
                    position: (Position(1, 0), Position(4, 1)),
                    len: 14,
                    level: 1
                }
            );
        }

        #[test]
        fn it_should_error_when_content_block_is_not_closed_properly() {
            let mut stream = tokenize("`\n\tabc`");
            match stream.next().unwrap() {
                Ok(_) => panic!("parsed invalid content"),
                Err(e) => assert!(match e {
                    LexerError::UnexpectedEndOfInput(_) => true,
                    _ => false
                })
            }
        }
    }

    mod single_char_tokens {
        use super::{tokenize, Position, RaToken, TokenKind};
        #[test]
        fn it_should_parse_tokens() {
            let mut stream = tokenize("!?&/#");
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Exclamation,
                    position: (Position(1, 0), Position(1, 1)),
                    content: "!",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Question,
                    position: (Position(1, 1), Position(1, 2)),
                    content: "?",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Ampersand,
                    position: (Position(1, 2), Position(1, 3)),
                    content: "&",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::Slash,
                    position: (Position(1, 3), Position(1, 4)),
                    content: "/",
                    level: 1,
                    len: 1
                }
            );
            assert_eq!(
                stream.next().unwrap().unwrap(),
                RaToken {
                    kind: TokenKind::HashPound,
                    position: (Position(1, 4), Position(1, 5)),
                    content: "#",
                    level: 1,
                    len: 1
                }
            );
        }
    }


}
