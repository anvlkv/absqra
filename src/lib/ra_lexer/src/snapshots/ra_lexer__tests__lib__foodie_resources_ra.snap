---
source: src/lib/ra_lexer/src/tests.rs
expression: tokens
---
[
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        1,
        0
      ],
      [
        1,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "navigationItems"
    },
    "len": 15,
    "content": "navigationItems",
    "position": [
      [
        1,
        1
      ],
      [
        1,
        16
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "meals"
    },
    "len": 5,
    "content": "meals",
    "position": [
      [
        2,
        0
      ],
      [
        2,
        5
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        3,
        4
      ],
      [
        3,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        3,
        5
      ],
      [
        3,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        3,
        7
      ],
      [
        3,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        3,
        8
      ],
      [
        3,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 29,
    "content": "\n            Meals\n        ",
    "position": [
      [
        3,
        10
      ],
      [
        5,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        6,
        4
      ],
      [
        6,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        6,
        5
      ],
      [
        6,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        6,
        7
      ],
      [
        6,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        6,
        8
      ],
      [
        6,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 34,
    "content": "\n            Блюда\n        ",
    "position": [
      [
        6,
        10
      ],
      [
        8,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "countries"
    },
    "len": 9,
    "content": "countries",
    "position": [
      [
        9,
        0
      ],
      [
        9,
        9
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        10,
        4
      ],
      [
        10,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        10,
        5
      ],
      [
        10,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        10,
        7
      ],
      [
        10,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        10,
        8
      ],
      [
        10,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 33,
    "content": "\n            Countries\n        ",
    "position": [
      [
        10,
        10
      ],
      [
        12,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        13,
        4
      ],
      [
        13,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        13,
        5
      ],
      [
        13,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        13,
        7
      ],
      [
        13,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        13,
        8
      ],
      [
        13,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 59,
    "content": "\n            Национальные кухни\n        ",
    "position": [
      [
        13,
        10
      ],
      [
        15,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "restrictions"
    },
    "len": 12,
    "content": "restrictions",
    "position": [
      [
        16,
        0
      ],
      [
        16,
        12
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        17,
        4
      ],
      [
        17,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        17,
        5
      ],
      [
        17,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        17,
        7
      ],
      [
        17,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        17,
        8
      ],
      [
        17,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 35,
    "content": "\n            Preferences\n        ",
    "position": [
      [
        17,
        10
      ],
      [
        19,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        20,
        4
      ],
      [
        20,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        20,
        5
      ],
      [
        20,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        20,
        7
      ],
      [
        20,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        20,
        8
      ],
      [
        20,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 48,
    "content": "\n            Предпочтения\n        ",
    "position": [
      [
        20,
        10
      ],
      [
        22,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        24,
        0
      ],
      [
        24,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "userItems"
    },
    "len": 9,
    "content": "userItems",
    "position": [
      [
        24,
        1
      ],
      [
        24,
        10
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "addRecipe"
    },
    "len": 9,
    "content": "addRecipe",
    "position": [
      [
        25,
        0
      ],
      [
        25,
        9
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        26,
        4
      ],
      [
        26,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "kind"
    },
    "len": 4,
    "content": "kind",
    "position": [
      [
        26,
        5
      ],
      [
        26,
        9
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        26,
        9
      ],
      [
        26,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "image"
    },
    "len": 5,
    "content": "image",
    "position": [
      [
        26,
        10
      ],
      [
        26,
        15
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 44,
    "content": "\n            http://path.to.image\n        ",
    "position": [
      [
        26,
        15
      ],
      [
        28,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": "Coma",
    "len": 1,
    "content": ",",
    "position": [
      [
        28,
        5
      ],
      [
        28,
        6
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        29,
        4
      ],
      [
        29,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        29,
        5
      ],
      [
        29,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        29,
        7
      ],
      [
        29,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        29,
        8
      ],
      [
        29,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 34,
    "content": "\n            Add recipe\n        ",
    "position": [
      [
        29,
        10
      ],
      [
        31,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        32,
        4
      ],
      [
        32,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        32,
        5
      ],
      [
        32,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        32,
        7
      ],
      [
        32,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        32,
        8
      ],
      [
        32,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 53,
    "content": "\n            Добавить рецепт\n        ",
    "position": [
      [
        32,
        10
      ],
      [
        34,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "myAccount"
    },
    "len": 9,
    "content": "myAccount",
    "position": [
      [
        35,
        0
      ],
      [
        35,
        9
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        36,
        4
      ],
      [
        36,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        36,
        5
      ],
      [
        36,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        36,
        7
      ],
      [
        36,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        36,
        8
      ],
      [
        36,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 34,
    "content": "\n            My account\n        ",
    "position": [
      [
        36,
        10
      ],
      [
        38,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        39,
        4
      ],
      [
        39,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        39,
        5
      ],
      [
        39,
        7
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        39,
        7
      ],
      [
        39,
        8
      ]
    ],
    "level": 2
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        39,
        8
      ],
      [
        39,
        10
      ]
    ],
    "level": 2
  },
  {
    "kind": "ContentBlock",
    "len": 45,
    "content": "\n            Мой аккаунт\n        ",
    "position": [
      [
        39,
        10
      ],
      [
        41,
        5
      ]
    ],
    "level": 2
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        43,
        0
      ],
      [
        43,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "searchPlaceholder"
    },
    "len": 17,
    "content": "searchPlaceholder",
    "position": [
      [
        43,
        1
      ],
      [
        43,
        18
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        44,
        0
      ],
      [
        44,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        44,
        1
      ],
      [
        44,
        3
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        44,
        3
      ],
      [
        44,
        4
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        44,
        4
      ],
      [
        44,
        6
      ]
    ],
    "level": 1
  },
  {
    "kind": "ContentBlock",
    "len": 29,
    "content": "\n        Find a recipe\n    ",
    "position": [
      [
        44,
        6
      ],
      [
        46,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        47,
        0
      ],
      [
        47,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        47,
        1
      ],
      [
        47,
        3
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        47,
        3
      ],
      [
        47,
        4
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        47,
        4
      ],
      [
        47,
        6
      ]
    ],
    "level": 1
  },
  {
    "kind": "ContentBlock",
    "len": 39,
    "content": "\n        Найти рецепт\n    ",
    "position": [
      [
        47,
        6
      ],
      [
        49,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        51,
        0
      ],
      [
        51,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "pictureLabel"
    },
    "len": 12,
    "content": "pictureLabel",
    "position": [
      [
        51,
        1
      ],
      [
        51,
        13
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        52,
        0
      ],
      [
        52,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        52,
        1
      ],
      [
        52,
        3
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        52,
        3
      ],
      [
        52,
        4
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "en"
    },
    "len": 2,
    "content": "en",
    "position": [
      [
        52,
        4
      ],
      [
        52,
        6
      ]
    ],
    "level": 1
  },
  {
    "kind": "ContentBlock",
    "len": 23,
    "content": "\n        Picture\n    ",
    "position": [
      [
        52,
        6
      ],
      [
        54,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": "HashPound",
    "len": 1,
    "content": "#",
    "position": [
      [
        55,
        0
      ],
      [
        55,
        1
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ln"
    },
    "len": 2,
    "content": "ln",
    "position": [
      [
        55,
        1
      ],
      [
        55,
        3
      ]
    ],
    "level": 1
  },
  {
    "kind": "Colon",
    "len": 1,
    "content": ":",
    "position": [
      [
        55,
        3
      ],
      [
        55,
        4
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Identifier": "ru"
    },
    "len": 2,
    "content": "ru",
    "position": [
      [
        55,
        4
      ],
      [
        55,
        6
      ]
    ],
    "level": 1
  },
  {
    "kind": "ContentBlock",
    "len": 32,
    "content": "\n        Картинка\n    ",
    "position": [
      [
        55,
        6
      ],
      [
        57,
        1
      ]
    ],
    "level": 1
  }
]
